packages/create-emotion-server/types/index.d.ts

```diff
@@ -1,12 +1,19 @@
+// Definitions by: Junyoung Clare Jang <https://github.com/Ailrun>
+// TypeScript Version: 2.8
+
 /// <reference types="node" />
-// TypeScript Version: 2.3
+import { EmotionCache } from 'emotion'
 
-import { Emotion } from 'create-emotion';
+export interface EmotionCritical {
```

Comment:
Not a fan of this, why would someone want to use this interface?

Reply:
When someone want to type everything. For example, in following case.
```
const { html, ids, css }: EmotionCritical = emotionServer.extractCritical(renderedString);
const { html, ids, css }: { html: string; ids: Array<string>; css: string } = emotionServer.extractCritical(renderedString);
```
Second one is definitely more complex, IMO.

Reply:
But why would anyone want to do that?

Reply:
It’s just more effort for no gain in type safety

Reply:
@mitchellhamilton 
In TS, enforcing such rule does give some type safety sometimes, since TS does not infer type well just as flow does.
(Sometimes it infer weird, too wide, or too narrow type since it (usually) infers type not based on environment, but just based on the single statement)

Reply:
@mitchellhamilton Actually, my team does 😃

Reply:
In this case, how would TS fail? I strongly disagree with that linting rule and I don’t think our types should be helping people do something which doesn’t actually help type safety and just adds more effort and makes the public API larger.

Reply:
Of course it does not help only for this case.
However, for the rule generally, I think you cannot understand the rule since your background is flow. Flow has context-based type inference (which is muuuuuuuuuuch better than TS' type inference), but TS does not, so, in Flow,
```js
declare function fun(arg: 0 | 1 | 2): void;
let x = 2;
fun(x);
```
will just work:tm:. However, in TS, [similar example](https://www.typescriptlang.org/play/index.html#src=declare%20function%20fun(arg%3A%200%20%7C%201%20%7C%202)%3A%20void%3B%0D%0Alet%20x%20%3D%202%3B%0D%0Afun(x)%3B%0D%0Alet%20y%3A%202%20%3D%202%3B%0D%0Afun(y)%3B%0D%0A) does not work, where explicitly typed one works.

Reply:
There are problem related with type generics too, but I cannot immediately create an example that works in flow but does not work in TS. (Usually, type generics case that does not work with TS is too complex to consider it as an example)